src/seq_block_mv/dense_block_matrix.c

```diff
@@ -0,0 +1,334 @@
+/******************************************************************************
+ * Copyright (c) 1998 Lawrence Livermore National Security, LLC and other
+ * HYPRE Project Developers. See the top-level COPYRIGHT file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ ******************************************************************************/
+
+/******************************************************************************
+ *
+ * Member functions for hypre_DenseBlockMatrix class.
+ *
+ *****************************************************************************/
+
+#include "_hypre_seq_block_mv.h"
+
+/*--------------------------------------------------------------------------
+ * hypre_DenseBlockMatrixCreate
+ *--------------------------------------------------------------------------*/
+
+hypre_DenseBlockMatrix *
+hypre_DenseBlockMatrixCreate( HYPRE_Int  row_major,
+                              HYPRE_Int  num_rows,
+                              HYPRE_Int  num_cols,
+                              HYPRE_Int  num_rows_block,
+                              HYPRE_Int  num_cols_block )
+{
+   hypre_DenseBlockMatrix  *A;
+   HYPRE_Int                num_blocks[2];
+
+   /* Compute number of blocks */
+   num_blocks[0] = 1 + ((num_rows - 1) / num_rows_block);
+   num_blocks[1] = 1 + ((num_cols - 1) / num_cols_block);
+   if (num_blocks[0] != num_blocks[1])
```

Comment:
You want this matrix block structure to be square? But you allow non-square dense blocks?
Is this needed?

Reply:
I just don't understand why you don't allow that the number of blocks in x-direction are not equal to the number of blocks in y-direction.

Reply:
Because it you think in 2D, there's actually just one block that spans both dimensions.

Suppose the full matrix is 8x8 and you are trying to build it with 2x4 local blocks. That shouldn't work because there isn't enough space in the second dimension. The check prevents a situation like this

Reply:
That is just a matter of definition. I don't see an issue there. I am thinking about a much more general matrix than you apparently. It seems that you want the original matrix to be square.

Reply:
No, the original matrix can be rectangular. But that implies the local blocks to be rectangular as well.

What I don't support is a mix of both, i.e., full matrix square while local block matrices rectangular (or vice-versa)

Reply:
So, you don't want to deal with interpolation matrices that might have the same block sizes? Those would not be allowed with that condition.

Reply:
Yes, that isn't supported here. I guess if we need this scenario to work, we could revisit.